### ISSUE TITLE
Slider on SwiftUi KSVideoPlayerView

### ISSUE THREAD

**Issue by UnknownCoder807** [kingslay/KSPlayer#232]
Using TVOS

At the moment to use the slider when using KSVideoPlayerView you have to press the ok/select button on the remote.  Then you can slide left and right.  You then need to press ok/select again to end.

Is it possible to have the slider get focus automatically and slide left and right without having to press ok/select first?  Pressing ok/select could still stop the slide and give focus back to navigate to other controls ?

so basically, give focus to the slider when you select it, without having to press ok/select first.

**Comment by UnknownCoder807** [comment]
I’ve found that the slider gets focus the first time it’s used.  If I move focus to a button then back to the slider, I need to press ok/select to make the slider work again.  Is it possible to allow slide left/right automatically whe the slider has focus?

**Comment by UnknownCoder807** [comment]
Also, the speed of the slide left/right is too fast.  Can it be slowed down half speed (for example) ?

**Comment by cdg90** [comment]
hello @Alanko5 I am asking you because you might know the solution.

I could not find how to forward/back forward +/- 10 seconds by pressing the left or right button, can you please tell me how to do it?

If it is not in KSplayer code, it would be cool to add the feature @kingslay

**Comment by Alanko5** [comment]
@cdg90 show me your code.
to see how you wrote it....

EXAMPLE:

register gesture:
```
            let menuPressRecognizer = UITapGestureRecognizer()
            menuPressRecognizer.addTarget(self, action: #selector(menuRightArrowAction(recognizer:)))
            menuPressRecognizer.allowedPressTypes = [NSNumber(value: UIPress.PressType.rightArrow.rawValue)]
            player.addGestureRecognizer(menuPressRecognizer)
```

forward action:
```
    @objc
    func menuRightArrowAction(recognizer:UITapGestureRecognizer) {
        let forwardToTime = player.player.currentTime().seconds + 10
        player.currentPlaybackTime = forwardToTime
    }
```

**Comment by cdg90** [comment]
> @cdg90 show me your code. to see how you wrote it....
> 
> EXAMPLE:
> 
> register gesture:
> 
> ```
>             let menuPressRecognizer = UITapGestureRecognizer()
>             menuPressRecognizer.addTarget(self, action: #selector(menuRightArrowAction(recognizer:)))
>             menuPressRecognizer.allowedPressTypes = [NSNumber(value: UIPress.PressType.rightArrow.rawValue)]
>             playerTopView.addGestureRecognizer(menuPressRecognizer)
> ```
> 
> forward action:
> 
> ```
>     @objc
>     func menuRightArrowAction(recognizer:UITapGestureRecognizer) {
>         let forwardToTime = player.currentTime().seconds + 10
>         player.currentPlaybackTime = forwardToTime
>     }
> ```

The thing is, I want to implement it into the KSVideoPlayerView UI that kingslay made. Do you have any hint?
(look in KSPlayer/SwiftUI/KSVideoPlayer.swift)

**Comment by UnknownCoder807** [comment]
> Also, the speed of the slide left/right is too fast. Can it be slowed down half speed (for example) ?

and

> I’ve found that the slider gets focus the first time it’s used. If I move focus to a button then back to the slider, I need to 
> press ok/select to make the slider work again. Is it possible to allow slide left/right automatically whe the slider has focus?

Any thoughts on this @kingslay ?

**Comment by cdg90** [comment]
> Also, the speed of the slide left/right is too fast. Can it be slowed down half speed (for example) ?

hello @kingslay and @Alanko5 the video slider work fine but the slide is very fast, and hard to stop at the time I want when I watch a VOD.
how can we reduce the speed by half, or 3x slower the slide left/right?

Thank you guys

**Comment by kingslay** [comment]
> > Also, the speed of the slide left/right is too fast. Can it be slowed down half speed (for example) ?
> 
> and
> 
> > I’ve found that the slider gets focus the first time it’s used. If I move focus to a button then back to the slider, I need to
> > press ok/select to make the slider work again. Is it possible to allow slide left/right automatically whe the slider has focus?
> 
> Any thoughts on this @kingslay ?

if allow slide left/right automatically when the slider has focus. the slide can not move focus to next button. so i add a UITapGestureRecognizer to controls whether to swipe left and right

**Comment by UnknownCoder807** [comment]
> if allow slide left/right automatically when the slider has focus. the slide can not move focus to next button. so i add a UITapGestureRecognizer to controls whether to swipe left and right

I understand, makes sens.

Can you help with the slider speed?  it's way too fast and smaller slide ,movements are very difficult if not impossible.  It's not to bad on the simulator, but on a real remote it's more visible.

**Comment by kingslay** [comment]
```
// KSVideoPlayer.swift L646
 let touchPoint = sender.location(in: self)
        let value = Float(touchPoint.x / frame.size.width)
```
the slide speed is can not reduce 
if you  set 

`        let value = Float(touchPoint.x / frame.size.width) / 2
`
then you can not set the value to  1

i you can try it

**Comment by kingslay** [comment]
@UnknownCoder807 
I'm wondering if sliders are a good interaction in tvOS. Whether Picker would be a better way. Let the user choose a specific time

**Comment by UnknownCoder807** [comment]
I tried with appending / 2 to the line as indicated but it's not working correctly.  It's not possible to slide towards the end of the slider.  The max position seems to be about 70%.

**Comment by UnknownCoder807** [comment]
> @UnknownCoder807 I'm wondering if sliders are a good interaction in tvOS. Whether Picker would be a better way. Let the user choose a specific time

I'm not sure that would be user friendly.  Have you an ida on how it would work?

**Comment by kingslay** [comment]
> I tried with appending / 2 to the line as indicated but it's not working correctly. It's not possible to slide towards the end of the slider. The max position seems to be about 70%.
> 
> edit : if the current time could be constantly updated when you are sliding, it would help with getting to a specific time.

Yes, that's the question. That's why I said I can't reduce speed

**Comment by kingslay** [comment]
> > @UnknownCoder807 I'm wondering if sliders are a good interaction in tvOS. Whether Picker would be a better way. Let the user choose a specific time
> 
> I'm not sure that would be user friendly. Have you an idea on how it would work?

How do other video players let users choose  time?

**Comment by UnknownCoder807** [comment]
> > > @UnknownCoder807 I'm wondering if sliders are a good interaction in tvOS. Whether Picker would be a better way. Let the user choose a specific time
> > 
> > 
> > I'm not sure that would be user friendly. Have you an idea on how it would work?
> 
> How do other video players let users choose time? I don't have tvos on my side, so I don't know how others do it

They do it using a slider.  I presume they use something like https://github.com/zattoo/TvOSSlider (which I use in another app with ksplayer)

https://github.com/zattoo/TvOSSlider is a port of the UIKit UISlider.  I did get it working with swiftui a while ago.

**Comment by Alanko5** [comment]
@UnknownCoder807 
Check me code:
https://github.com/Alanko5/KSPlayer/blob/develop/Sources/KSPlayer/Video/tvOS/Bar/ProgressToolBarView.swift

**Comment by kingslay** [comment]
@UnknownCoder807  Then you can push request to change the code of the TVSlide class. Change it to be similar to the interaction with zattoo/TvOSSlider

**Comment by UnknownCoder807** [comment]
@Alanko5 thanks for the information.  Sadly I'm not that confident on how to implement it into SwiftUI.  Thanks though for taking the time.

**Comment by UnknownCoder807** [comment]
@kingslay please look at this short screen capture video I made afer adding the " / 2" fix as above.

https://collect.wetransfer.com/board/spjgt24jqh43gkwx920220508162940/latest

When I slide right on the remote, the red progress bar jumps back to an earlier position and then starts to slide from that position.  What it should do is start sliding from the current position of the red progress bar.  Hope that makes sense.

Maybe a bug in the starting position for sliding... ?

**Comment by kingslay** [comment]
@UnknownCoder807 you can try use sender.translation(in: self)  (TvOSSlider use it) or sender.velocity(in: self)  ( @Alanko5  use it ) to update the value. I is use sender.location(in: self)

**Comment by UnknownCoder807** [comment]
> @UnknownCoder807 you can try use sender.translation(in: self) (TvOSSlider use it) or sender.velocity(in: self) ( @Alanko5 use it ) to update the value. I is use sender.location(in: self)

Using sender.translation or sender.velocity makes it unusable completely.

If I open a video for playing, and the current position is 50%. I then use the slider to fast forward. The slider slides from 50% to 75% (for example) (which is correct).  If I then slide again to the left, the slider slides from 50%, not from 75%.  Can be seen in the video I posted above.

**Comment by UnknownCoder807** [comment]
I found a solution although I'm not sure it's very elgant.

```
        switch sender.state {
        case .began, .possible:
            isTouch = true
        case .changed:
            if translation.x < 0 { // user has swiped left
                // Increase 0.005 to speed up the slider
                // Decrease 0.005 to slow down the slider
                processView.progress -= 0.005
            } else { // user has swiped right
                // Increase 0.005 to speed up the slider
                // Decrease 0.005 to slow down the slider
                processView.progress += 0.005
            }
        case .ended:
            process.wrappedValue = processView.progress
            isTouch = false
        case .cancelled:
            isTouch = false
        case .failed:
            isTouch = false
        @unknown default:
            break
        }
```

I don't know if there is a better way, but this way al least works.

**Comment by UnknownCoder807** [comment]
Closing, as don't think this will ever be fixed/working on tvOS properly.

**Comment by UnknownCoder807** [comment]
Thanks for the fix today @kingslay, it's a major upgrade in making the slider usable now.

There is however one small issue still.  When the user finishes sliding, the progress bar seems to jump very quickly to some position before going back to the correct end position.

This short video shows it.

https://user-images.githubusercontent.com/56231066/168464835-caa144f0-9208-4701-a167-9268f32e72c6.mov

**Comment by Alanko5** [comment]
This is wrong way.
But you must pause movie an then enable scrolling to position “seek”….

**Comment by UnknownCoder807** [comment]
I've never come across an App on Apple hardware that pauses the video, seek, then unpause.  Doesn't sound very intuitive to me.  But if this is the way it must be, so be it :)

**Comment by UnknownCoder807** [comment]
I think whats happening with the cosmetic graphical glitch (as shown in the video above), is when you stop swiping on the remote, the red bar temporarily jumps back to the start position and then jumps back to the end position.  Maybe this is fixable?

**Comment by Alanko5** [comment]
for example, HBO MAX.
click for fast rewind, swipe does not work during playback.

**Comment by UnknownCoder807** [comment]
> for example, HBO MAX. click for fast rewind, swipe does not work during playback.

It's been pointed out to me that the Amazon Prime App on the Apple TV also pauses when seeking.  I just tried it and I concede I was wrong, it does work ok.

So the question is, how to get access to the player from within the slider.  Is it possible ? Also, to make it work properly, the time the user is seeking to, needs to be displayed in real time.  @kingslay

**Comment by UnknownCoder807** [comment]
@kingslay any thoughts on my previous post?  I think what alanko says is the correct way to go.

**Comment by UnknownCoder807** [comment]
@kingslay thank you for the fixes.  The slider now works nicely.

**Comment by kingslay** [comment]
```
  case .ended:
            onEditingChanged(false)
        case .cancelled, .failed:
            process.wrappedValue = beganProgress
            onEditingChanged(false)
```
Could you please check if the gesture goes to ended or failed @UnknownCoder807

**Comment by UnknownCoder807** [comment]
> ```
>   case .ended:
>             onEditingChanged(false)
>         case .cancelled, .failed:
>             process.wrappedValue = beganProgress
>             onEditingChanged(false)
> ```
> 
> Could you please check if the gesture goes to ended or failed @UnknownCoder807

When the gesture doesn't work, the gesture goes to .ended

**Comment by kingslay** [comment]
```
  Slider(value: $model.currentTime, in: 0 ... model.totalTime) { onEditingChanged in
                    if onEditingChanged {
                        config.isPlay = false
                    } else {
                        config.seek(time: model.currentTime)
                    }
                }
```
Then you debug to see what is the value passed to config.seek

**Comment by UnknownCoder807** [comment]
The model.currentTime seems to be random numbers

-*-*-*-*-*-*-*-*-*-* 777.9990649819374 -*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* 354.751 -*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* 373.748 -*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* 1001.9281162619591 -*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* 398.751 -*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* 663.9591756314039 -*-*-*-*-*-*-*-*-*-*

This is sliding from the same start position to approx the same end position.

**Comment by UnknownCoder807** [comment]
If I slide right and stop, without lifting my finger off the remote, the slider resets back to the start position after a second or so.  don't know if this is relevant or not.

**Comment by UnknownCoder807** [comment]
Bit more logging.  I put a log in the .began, .changed and ended options.  I also put a log in the Slider seek code to see where it was seeking too.

```
-*-*-*-*-*-*-*-*-*-* began (Start = 0.08019771)-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.08158186-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.08158186-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.083558775-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.08479086-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.0855388-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.08676778-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.08751572-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.08826676-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.08949574-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* ended -*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* (Seek time = 564.1072042658925) -*-*-*-*-*-*-*-*-*-*
```

I don't know if that helps.

**Comment by UnknownCoder807** [comment]
I can send you a video url to try yourself if it helps.

**Comment by kingslay** [comment]
you need print the totalTime

**Comment by UnknownCoder807** [comment]
> you need print the totalTime

```
-*-*-*-*-*-*-*-*-*-* began (Start = 0.0011524827)-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.010143272-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.010143272-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.02024821-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.031641096-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.03831048-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.049703363-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.059808303-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.06991014-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.07765023-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.080067836-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.08482237-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.08872035-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.09114107-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.09237004-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.09311798-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.09386902-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.09420731-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* ended -*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* (total time = 8237.0) -*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* (current time = 775.9856057688594) -*-*-*-*-*-*-*-*-*-*
```

**Comment by kingslay** [comment]
If the seek time is 775.9856057688594. That's right

**Comment by UnknownCoder807** [comment]
When the slider works ok, there is this difference

```
-*-*-*-*-*-*-*-*-*-* changed (to 0.24045265-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* changed (to 0.24079093-*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* ended -*-*-*-*-*-*-*-*-*-*
KSPlayer: KSPlayerLayer.swift:144 state | playerStateDidChange - buffering
KSPlayer: KSMEPlayer.swift:305 play() | play <KSPlayer.KSMEPlayer: 0x600000b42080>
KSPlayer: KSPlayerLayer.swift:144 state | playerStateDidChange - bufferFinished
-*-*-*-*-*-*-*-*-*-* (total time = 8913.0) -*-*-*-*-*-*-*-*-*-*
-*-*-*-*-*-*-*-*-*-* (current time = 2146.1695891320705) -*-*-*-*-*-*-*-*-*-*
```

**Comment by kingslay** [comment]
I don't quite know what you mean. You need to look after the gesture is over. Whether the seek time is the same as the time you selected last. If it is the same, then there is no problem. You can print the value of each link if it is different. So it could be that link that makes the difference

**Comment by UnknownCoder807** [comment]
The seek time is correct.  It is the time that is selected.  The difference is, when the slider works, KSPlayer outputs

```
KSPlayer: KSPlayerLayer.swift:144 state | playerStateDidChange - buffering
KSPlayer: KSMEPlayer.swift:305 play() | play <KSPlayer.KSMEPlayer: 0x600000b42080>
KSPlayer: KSPlayerLayer.swift:144 state | playerStateDidChange - bufferFinished
```
which means the seek actually seeks.

When the slider doesn't work, it doesn't display it.  So it's not actually seeking to the time selected.

**Comment by UnknownCoder807** [comment]
@kingslay have you any ideas on how I can debug this?  it's completely random.  sometimes the slider works on a video and then when I play the video again, it won't work.

**Comment by kingslay** [comment]
you can print the value to debug

**Comment by UnknownCoder807** [comment]
The value of what?  I have printed the values above.  the seek command is being called, but it's not completing for some reason.  I think debugging the seek command itself to see why its completing is worth doing ?

**Comment by UnknownCoder807** [comment]
I setup some break points from the config.seek(time: model.currentTime)

This calls this function 

```
        func seek(time: TimeInterval) {
            playerLayer?.seek(time: time, autoPlay: true)
        }
```

putting a break point on playerLayer?.seek

execution stops as expected but when I think step forward playerLayer?.seek is not called.

I printed the value of "time" and it has a current value.

**Comment by UnknownCoder807** [comment]
I put a breakpoint in playerLayer?.seek(time: time, autoPlay: true) and its not called when ending the slide.  So the slider itself is working.

```
                            Slider(value: $model.currentTime, in: 0 ... model.totalTime) { onEditingChanged in
                                if onEditingChanged {
                                    config.isPlay = false
                                } else {
                                    config.seek(time: model.currentTime)
                                }
                            }
                            .frame(maxHeight: 20)
```

config.seek calls 

```
        func seek(time: TimeInterval) {
            playerLayer?.seek(time: time, autoPlay: true)
        }
```

but playerLayer?.seek(time: time, autoPlay: true) is not called.

hope that explains it better.

**Comment by kingslay** [comment]
Then this is weird. How can config.seek(time: model.currentTime) be called. But the func seek(time: TimeInterval) is not executed

**Comment by UnknownCoder807** [comment]
I don't know if this is relevant, but if I put a breakpoint into 

```
        func seek(time: TimeInterval) {
            playerLayer?.seek(time: time, autoPlay: true)
        }
```
and then 

`po playerLayer 
`

in the debug window in xcode, it returns nil

![tmp](https://user-images.githubusercontent.com/56231066/170450889-672ced76-51e3-4deb-aed4-ebfe0327178e.jpg)

Could this explain why the seek is not run ?

**Comment by kingslay** [comment]
It is not expected that playerLayer is nil. You need to debug it. When did the variable playerLayer become nil?

**Comment by UnknownCoder807** [comment]
I've done some debugging.

I have output the value of playerLayer is numerous places and it always has a value except for one place.

if I output the value of playerView inside 

```
        public func player(layer _: KSPlayerLayer, currentTime: TimeInterval, totalTime: TimeInterval) {
            onPlay?(currentTime, totalTime)
        }
```

playerLayer has a value all the time even when the slider doesn't work.

When the slider doesn't work, the following swipe right action still works, so config.seek is working, just not inside the slider.

`config.seek(time: model.currentTime + 15)
`

All I can work out is

```
        func seek(time: TimeInterval) {
            playerLayer?.seek(time: time, autoPlay: true)
        }
```
playerLayer returns nil here.


Is there some specific debugging you think I should try?  

The slider works some of the time, using the same video file, it can work, then not work, then work etc

**Comment by kingslay** [comment]
```

        fileprivate weak var playerLayer: KSPlayerLayer? {
            willSet {
                if newValue == nil {
                    print("")
                }
            }
        }
```

**Comment by UnknownCoder807** [comment]
Thanks, tried the above and it didn't output anything, so playerlayer is being set.

**Comment by UnknownCoder807** [comment]
Just for closure on this issue, I think I found the problem.  

It's an issue with having multiple NavigationLink() on a screen.  SwiftUI is calling the init() for each instance of NavigationLink() when the screen is loaded.  By stopping init() from being called, the slider issue went away.  Strange.
